compiler/rustc_mir/src/transform/deduplicate_blocks.rs

```diff
@@ -0,0 +1,233 @@
+        opt_applier.visit_body(body);
+
+        if should_cleanup {
+            simplify_cfg(body);
+        }
+    }
+}
+
+fn find_duplicates<'a, 'tcx>(body: &'a Body<'tcx>) -> FxHashMap<BasicBlock, BasicBlock> {
+    let mut duplicates = FxHashMap::default();
+
+    // We want to avoid O(n²) complexity by comparing every basic block with each other.
+    // Basic blocks are equal iff
+    // - their length of statements are equal
+    // - their TerminatorKind are equal
+    // - their StatementKinds are pairwise equal
+    //
+    // The first three cases are cheap to checks, so we use the following strategy:
+    // 1. Sort the basic blocks by (length of statements, TerminatorKind)
+    //      - This is needed for itertools::group_by which only assigns consecutive elements to the same group
+    // 2. Group by (length of statements, TerminatorKind)
+    // 3. Compare StatementKinds pairwise inside the group
+    //      - This is technically O(n²), but `n` should be small in most cases
+
+    let mut to_sort: Vec<_> = body
+        .basic_blocks()
+        .iter_enumerated()
+        // Fixme(simonvandel): If we find is_cleanup = true, the compiler ICEs in library/std/src/sys/unix/process/process_unix.rs exec
+        // with the error `funclet bb11 has 2 parents - bb24 and bb26`
+        .filter(|(_, bb_data)| !bb_data.is_cleanup)
+        .map(|(bb_idx, bb_data)| {
+            (
+                bb_idx,
+                Metadata {
+                    len_statements: bb_data.statements.len(),
+                    terminator_kind: bb_data.terminator().kind.clone(),
+                },
+            )
+        })
+        .collect();
+
+    // We do not use an unstable sort, as we want to preserve the order of the basic blocks.
+    // The current order is always increasing. This property is important,
+    // as we are using the same order when replacing basic blocks in terminators.
+    // With an ascending basic block index order, if bb1, bb2, and bb3 were all equal,
+    // we would replace bb1 with bb2, and then bb2 with bb3.
+    // If we used an unstable sort, we could potentially replace bb2 with bb3, and then bb1 with bb2.
+    // This would leave both bb3 and bb2, thereby not getting the full potential of this optimization.
+    to_sort.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
+    let iter = &to_sort.into_iter().group_by(|x| x.1.clone());
+    for x in iter {
+        let (group_key, group): (_, _) = x;
+        // We know by this point that the length of statements,
+        // and the TerminatorKind are equal within the group.
+        // Check that the statements are pairwise equal
+        let group: Vec<_> = group.map(|x| x.0).collect();
+        debug!("group: `{:?}` ", group);
+
+        let mut tuple_iter = TupleCombinationsWithSkip::new(group);
+        debug!("tuple_iter: {:?}", tuple_iter);
+        while let Some((item1, item2)) = tuple_iter.next() {
```

Comment:
Surely the most common case are `(len statements, terminator kind)` groups of size one, right? Could it be beneficial to filter them out early, since there obviously can't be any duplicates?

Reply:
Yep, I have added a filtering group size > 1 in the latest commits